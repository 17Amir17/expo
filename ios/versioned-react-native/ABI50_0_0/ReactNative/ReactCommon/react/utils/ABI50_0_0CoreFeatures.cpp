/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "ABI50_0_0CoreFeatures.h"

namespace ABI50_0_0facebook::ABI50_0_0React {

bool CoreFeatures::enablePropIteratorSetter = false;
bool CoreFeatures::blockPaintForUseLayoutEffect = false;
bool CoreFeatures::useNativeState = false;
bool CoreFeatures::cacheLastTextMeasurement = false;
bool CoreFeatures::cancelImageDownloadsOnRecycle = false;
bool CoreFeatures::enableGranularScrollViewStateUpdatesIOS = false;
bool CoreFeatures::enableMountHooks = false;
bool CoreFeatures::doNotSwapLeftAndRightOnAndroidInLTR = false;
bool CoreFeatures::enableCleanParagraphYogaNode = false;
bool CoreFeatures::disableScrollEventThrottleRequirement = false;
bool CoreFeatures::enableGranularShadowTreeStateReconciliation = false;
bool CoreFeatures::enableDefaultAsyncBatchedPriority = false;
bool CoreFeatures::enableClonelessStateProgression = false;
bool CoreFeatures::excludeYogaFromRawProps = false;

} // namespace ABI50_0_0facebook::ABI50_0_0React
