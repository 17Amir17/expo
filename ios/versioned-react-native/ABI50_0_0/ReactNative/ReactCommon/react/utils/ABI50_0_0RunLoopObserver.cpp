/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "ABI50_0_0RunLoopObserver.h"

#include <ABI50_0_0React/debug/ABI50_0_0React_native_assert.h>

namespace ABI50_0_0facebook::ABI50_0_0React {

RunLoopObserver::RunLoopObserver(
    Activity activities,
    const WeakOwner& owner) noexcept
    : activities_(activities), owner_(owner) {}

void RunLoopObserver::setDelegate(const Delegate* delegate) const noexcept {
  // We need these constraints to ensure basic thread-safety.
  ABI50_0_0React_native_assert(delegate && "A delegate must not be `nullptr`.");
  ABI50_0_0React_native_assert(
      !delegate_ && "`RunLoopObserver::setDelegate` must be called once.");
  delegate_ = delegate;
}

void RunLoopObserver::enable() const noexcept {
  if (enabled_) {
    return;
  }
  enabled_ = true;

  startObserving();
}

void RunLoopObserver::disable() const noexcept {
  if (!enabled_) {
    return;
  }
  enabled_ = false;

  stopObserving();
}

void RunLoopObserver::activityDidChange(Activity activity) const noexcept {
  if (!enabled_) {
    return;
  }

  ABI50_0_0React_native_assert(
      !owner_.expired() &&
      "`owner_` is null. The caller must `lock` the owner and check it for being not null.");

  delegate_->activityDidChange(delegate_, activity);
}

RunLoopObserver::WeakOwner RunLoopObserver::getOwner() const noexcept {
  return owner_;
}

} // namespace ABI50_0_0facebook::ABI50_0_0React
