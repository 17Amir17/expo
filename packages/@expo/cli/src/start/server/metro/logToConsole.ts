/**
 * Copyright Â© 2023 650 Industries.
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree. *
 */
import chalk from 'chalk';
import type { Terminal } from 'metro-core';
import util from 'util';

import { getMetroStackAsLogString, getSymbolicatedMetroStackAsync } from './metroErrorInterface';

const groupStack: ClientLogLevel[] = [];
let collapsedGuardTimer: ReturnType<typeof setTimeout>;

type ClientLogLevel =
  | 'error'
  | 'trace'
  | 'info'
  | 'warn'
  | 'log'
  | 'group'
  | 'groupCollapsed'
  | 'groupEnd'
  | 'debug';

type ClientLogMode = 'BRIDGE' | 'NOBRIDGE';

function colorFunctionForLevel(level: ClientLogLevel) {
  return level === 'error'
    ? chalk.inverse.red
    : level === 'warn'
    ? chalk.inverse.yellow
    : chalk.inverse.white;
}

async function logToConsoleAsync(
  projectRoot: string,
  terminal: Terminal,
  level: ClientLogLevel,
  mode: ClientLogMode,
  ...data: any[]
) {
  const logFunction = console[level] && level !== 'trace' ? level : 'log';
  const color = colorFunctionForLevel(level);

  if (level === 'group') {
    groupStack.push(level);
  } else if (level === 'groupCollapsed') {
    groupStack.push(level);
    clearTimeout(collapsedGuardTimer);
    // Inform users that logs get swallowed if they forget to call `groupEnd`.
    collapsedGuardTimer = setTimeout(() => {
      if (groupStack.includes('groupCollapsed')) {
        terminal.log(
          chalk.inverse.yellow.bold(' WARN '),
          'Expected `console.groupEnd` to be called after `console.groupCollapsed`.'
        );
        groupStack.length = 0;
      }
    }, 3000);
    return;
  } else if (level === 'groupEnd') {
    groupStack.pop();
    if (!groupStack.length) {
      clearTimeout(collapsedGuardTimer);
    }
    return;
  }

  if (groupStack.includes('groupCollapsed')) {
    return;
  }

  // Remove excess whitespace at the end of a log message, if possible.
  const lastItem = data[data.length - 1];

  if (typeof lastItem === 'string') {
    data[data.length - 1] = lastItem.trimEnd();
  }

  const modePrefix = !mode || mode === 'BRIDGE' ? '' : `(${mode.toUpperCase()}) `;

  // TODO: Can we add the platform tag back? Maybe there's a hint in the websocket connection? Otherwise we could pull from the stack trace sometimes
  // finally, we could try to PR support into HMRClient in React Native.

  const tMode =
    color.bold(` ${modePrefix}${logFunction.toUpperCase()} `) +
    ''.padEnd(groupStack.length * 2, ' ');

  // NOTE:
  // - If there's a stack trace, then symbolicate and dim based on the Expo Metro config `INTERNAL_CALLSITES_REGEX`
  //   console.log('data', data);

  if (['error', 'warn'].includes(level) && data.length > 1) {
    // console.log(data[1]);
    const results = await getSymbolicatedMetroStackAsync(projectRoot, {
      message: data[0],
      stack: data[1].match(/^\s+at\s((?:\n|.*)+)/im)?.[0],
    });

    if (results) {
      terminal.log(
        tMode + ' ' + data[0] + '\n',
        '\n' + getMetroStackAsLogString(projectRoot, results)
      );
      return;
    }
  }

  terminal.log(
    tMode,
    // `util.format` actually accepts any arguments.
    // If the first argument is a string, it tries to format it.
    // Otherwise, it just concatenates all arguments.
    // $FlowIssue[incompatible-call] util.format expected the first argument to be a string
    util.format(...data)
  );
}

export default logToConsoleAsync;
