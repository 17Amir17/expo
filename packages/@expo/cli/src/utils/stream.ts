function decodeStream(textDecoder = new TextDecoder()) {
  return new TransformStream<Uint8Array, string>({
    transform(chunk, controller) {
      return controller.enqueue(textDecoder.decode(chunk, { stream: true }));
    },
    flush(controller) {
      return controller.enqueue(textDecoder.decode());
    },
  });
}

// const streamToString = async (stream: ReadableStream): Promise<string> => {
//     const decoder = new TextDecoder();
//     const reader = stream.getReader();
//     const outs: string[] = [];
//     let result: ReadableStreamReadResult<unknown>;
//     do {
//       result = await reader.read();
//       if (result.value) {
//         if (!(result.value instanceof Uint8Array)) {
//           throw new Error('Unexepected buffer type');
//         }
//         outs.push(decoder.decode(result.value, { stream: true }));
//       }
//     } while (!result.done);
//     outs.push(decoder.decode());
//     return outs.join('');
//   };
export async function streamToStringAsync(stream: ReadableStream<Uint8Array>) {
  let output = '';

  await stream
    // Decode the streamed chunks to turn them into strings.
    .pipeThrough(decodeStream())
    .pipeTo(
      new WritableStream<string>({
        write(chunk) {
          output += chunk;
        },
      })
    );

  return output;
}
