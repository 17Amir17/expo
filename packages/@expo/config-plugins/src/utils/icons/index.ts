export * as AssetContents from './AssetContents';
export * as AndroidIcons from './withAndroidIcons';
export * as AndroidManifestIcons from './withAndroidManifestIcons';
export * as IosIcons from './withIosIcons';
