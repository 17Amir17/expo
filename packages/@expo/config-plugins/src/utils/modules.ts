import fs from 'fs';

const debug = require('debug')('expo:config-plugins:fs') as typeof console.log;

/**
 * A non-failing version of async FS stat.
 *
 * @param file
 */
async function statAsync(file: string): Promise<fs.Stats | null> {
  try {
    return await fs.promises.stat(file);
  } catch {
    return null;
  }
}

export async function fileExistsAsync(file: string): Promise<boolean> {
  return (await statAsync(file))?.isFile() ?? false;
}

export async function directoryExistsAsync(file: string): Promise<boolean> {
  return (await statAsync(file))?.isDirectory() ?? false;
}

export function fileExists(file: string): boolean {
  try {
    return fs.statSync(file).isFile();
  } catch {
    return false;
  }
}

// An optimization to attempt to prevent Xcode cache invalidation on files that don't change.
export async function writeIfDifferentAsync(filePath: string, contents: string): Promise<void> {
  if (!fileExists(filePath)) {
    const existing = await fs.promises.readFile(filePath, 'utf8');
    if (existing === contents) {
      debug(`Skipping writing unchanged file: ${filePath}`);
      return;
    }
  }
  await fs.promises.writeFile(filePath, contents);
}
