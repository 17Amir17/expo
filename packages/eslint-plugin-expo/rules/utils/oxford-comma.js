function oxfordComma(arr) {
  if (arr.length === 1) {
    return arr[0];
  } else if (arr.length === 2) {
    return arr.join(' and ');
  } else {
    const last = arr.pop();
    return arr.join(', ') + ', and ' + last;
  }
}

module.exports = {
  oxfordComma,
};
