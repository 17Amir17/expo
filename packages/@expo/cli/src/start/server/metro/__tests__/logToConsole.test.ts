import { Terminal } from 'metro-core';
import path from 'path';

import { stripAnsi } from '../../../../utils/ansi';
import logToConsole from '../logToConsole';

jest.unmock('resolve-from');

beforeEach(() => {
  delete process.env.EXPO_DEV_SERVER_ORIGIN;
});

afterAll(() => {
  delete process.env.EXPO_DEV_SERVER_ORIGIN;
});

jest.mock('../metroErrorInterface', () => ({
  getMetroStackAsLogString: jest.requireActual('../metroErrorInterface').getMetroStackAsLogString,
  getSymbolicatedMetroStackAsync: jest.fn(async () => ({
    stack: [
      {
        arguments: [],
        column: 2,
        file: '/my-app/node_modules/@react-navigation/native-stack/src/views/HeaderConfig.tsx',
        lineNumber: 35,
        methodName: 'HeaderConfig',
        collapse: false,
      },
      {
        arguments: [],
        column: 5,
        file: '/my-app/node_modules/react-native/Libraries/Animated/createAnimatedComponent.js',
        lineNumber: 39,
        methodName: 'React.forwardRef$argument_0',
        collapse: false,
      },
      {
        arguments: [],
        column: 10,
        file: '/my-app/node_modules/expo-router/src/ExpoRoot.tsx',
        lineNumber: 55,
        methodName: 'ExpoRoot',
        collapse: false,
      },
      {
        arguments: [],
        column: 26,
        file: '/my-app/node_modules/expo/build/launch/withDevTools.js',
        lineNumber: 31,
        methodName: 'WithDevTools',
        collapse: false,
      },
      {
        arguments: [],
        column: 0,
        file: '/my-app/node_modules/@expo/metro-runtime/build/error-overlay/toast/ErrorToastContainer.js',
        lineNumber: 9,
        methodName: 'ErrorToastContainer',
        collapse: true,
      },
      {
        arguments: [],
        column: 6,
        file: '/my-app/node_modules/react-native/Libraries/Components/View/View.js',
        lineNumber: 36,
        methodName: 'React.forwardRef$argument_0',
        collapse: false,
      },
      {
        arguments: [],
        column: 6,
        file: '/my-app/node_modules/react-native/Libraries/Components/View/View.js',
        lineNumber: 36,
        methodName: 'React.forwardRef$argument_0',
        collapse: false,
      },
      {
        arguments: [],
        column: 18,
        file: '/my-app/node_modules/react-native/Libraries/ReactNative/AppContainer.js',
        lineNumber: 40,
        methodName: 'AppContainer',
        collapse: false,
      },
      {
        arguments: [],
        column: 79,
        file: '/my-app/node_modules/react-native/Libraries/ReactNative/getCachedComponentWithDebugName.js',
        lineNumber: 28,
        methodName: 'getCachedComponentWithDisplayName',
        collapse: false,
      },
    ],
    codeFrame: {
      content:
        '\x1B[0m \x1B[90m 33 |\x1B[39m   headerHeight\x1B[33m,\x1B[39m\x1B[0m\n' +
        '\x1B[0m \x1B[90m 34 |\x1B[39m   headerBackImageSource\x1B[33m,\x1B[39m\x1B[0m\n' +
        '\x1B[0m\x1B[31m\x1B[1m>\x1B[22m\x1B[39m\x1B[90m 35 |\x1B[39m   headerBackButtonMenuEnabled\x1B[33m,\x1B[39m\x1B[0m\n' +
        '\x1B[0m \x1B[90m    |\x1B[39m   \x1B[31m\x1B[1m^\x1B[22m\x1B[39m\x1B[0m\n' +
        '\x1B[0m \x1B[90m 36 |\x1B[39m   headerBackTitle\x1B[33m,\x1B[39m\x1B[0m\n' +
        '\x1B[0m \x1B[90m 37 |\x1B[39m   headerBackTitleStyle\x1B[33m,\x1B[39m\x1B[0m\n' +
        '\x1B[0m \x1B[90m 38 |\x1B[39m   headerBackTitleVisible \x1B[33m=\x1B[39m \x1B[36mtrue\x1B[39m\x1B[33m,\x1B[39m\x1B[0m',
      location: { row: 35, column: 2 },
      fileName: '/my-app/node_modules/@react-navigation/native-stack/src/views/HeaderConfig.tsx',
    },
  })),
}));

const projectRoot = path.join(__dirname, '../../../../../../../../apps/router-e2e');

function createTerminal() {
  return {
    log: jest.fn(),
    // log: jest.fn(console.log),
  } as unknown as Terminal;
}

it(`symbolicates warning stack traces`, async () => {
  const terminal = createTerminal();
  await logToConsole(
    projectRoot,
    terminal,
    'warn',
    'BRIDGE',
    ...[
      'You started loading the font "Inter_500Medium", but used it before it finished loading. You need to wait for Font.loadAsync to complete before using the font.',
      '\n' +
        '    at HeaderConfig (http://172.20.1.239:8081/node_modules/expo-router/entry.bundle//&platform=ios&dev=true&hot=false&lazy=true:141762:28)\n' +
        '    at RNSScreen\n' +
        '    at anonymous (http://172.20.1.239:8081/node_modules/expo-router/entry.bundle//&platform=ios&dev=true&hot=false&lazy=true:98835:62)\n' +
        '    at Suspender (http://172.20.1.239:8081/node_modules/expo-router/entry.bundle//&platform=ios&dev=true&hot=false&lazy=true:141371:22)\n' +
        '    at Suspense\n' +
        '    at Freeze (http://172.20.1.239:8081/node_modules/expo-router/entry.bundle//&platform=ios&dev=true&hot=false&lazy=true:141390:23)\n' +
        '    at AppContainer (http://172.20.1.239:8081/node_modules/expo-router/entry.bundle//&platform=ios&dev=true&hot=false&lazy=true:40253:36)\n' +
        '    at main(RootComponent) (http://172.20.1.239:8081/node_modules/expo-router/entry.bundle//&platform=ios&dev=true&hot=false&lazy=true:121649:28)',
    ]
  );

  const res = terminal.log.mock.calls[0];
  expect(stripAnsi(res[1])).toMatchSnapshot();
});

it(`symbolicates console.error stack traces`, async () => {
  const terminal = createTerminal();
  await logToConsole(
    projectRoot,
    terminal,
    'error',
    'BRIDGE',
    ...[
      'hey',
      '\n' +
        '    at InnerLayout (http://172.20.1.239:8081/src/app/_layout.bundle//&platform=ios&hot=false&lazy=true&dev=true&minify=false&modulesOnly=true&runModule=false&shallow=true:102:48)\n' +
        '    at main(RootComponent) (http://172.20.1.239:8081/node_modules/expo-router/entry.bundle//&platform=ios&dev=true&hot=false&lazy=true:121649:28)',
    ]
  );

  const res = terminal.log.mock.calls[0];
  expect(stripAnsi(res[1])).toMatchSnapshot();
});

it(`doesn't symbolicate thrown errors`, async () => {
  const terminal = createTerminal();
  await logToConsole(
    projectRoot,
    terminal,
    'error',
    'BRIDGE',
    ...[
      'Error: thrown-in-tree\n' +
        '\n' +
        'This error is located at:\n' +
        '    in InnerLayout (created by Layout)\n' +
        '    in AnimatedSplashScreen (created by Layout)\n' +
        '    in Layout\n' +
        '    in Try\n' +
        '    in Unknown\n' +
        '    in RCTView (created by View)\n' +
        '    in View (created by AppContainer)\n' +
        '    in AppContainer\n' +
        '    in main(RootComponent), js engine: hermes',
    ]
  );

  const res = terminal.log.mock.calls[0].map(stripAnsi);
  expect(typeof res[0]).toBe('string');
  expect(typeof res[1]).toBe('string');
});
