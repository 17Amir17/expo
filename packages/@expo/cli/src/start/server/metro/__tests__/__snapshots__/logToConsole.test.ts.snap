// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`symbolicates console.error stack traces 1`] = `
"
  33 |   headerHeight,
  34 |   headerBackImageSource,
> 35 |   headerBackButtonMenuEnabled,
     |   ^
  36 |   headerBackTitle,
  37 |   headerBackTitleStyle,
  38 |   headerBackTitleVisible = true,

Call Stack
  HeaderConfig (my-app/node_modules/@react-navigation/native-stack/src/views/HeaderConfig.tsx:35:3)
  React.forwardRef$argument_0 (my-app/node_modules/react-native/Libraries/Animated/createAnimatedComponent.js:39:6)
  ExpoRoot (my-app/node_modules/expo-router/src/ExpoRoot.tsx:55:11)
  WithDevTools (my-app/node_modules/expo/build/launch/withDevTools.js:31:27)
  React.forwardRef$argument_0 (my-app/node_modules/react-native/Libraries/Components/View/View.js:36:7)
  React.forwardRef$argument_0 (my-app/node_modules/react-native/Libraries/Components/View/View.js:36:7)
  AppContainer (my-app/node_modules/react-native/Libraries/ReactNative/AppContainer.js:40:19)
  getCachedComponentWithDisplayName (my-app/node_modules/react-native/Libraries/ReactNative/getCachedComponentWithDebugName.js:28:80)
"
`;

exports[`symbolicates warning stack traces 1`] = `
"
  33 |   headerHeight,
  34 |   headerBackImageSource,
> 35 |   headerBackButtonMenuEnabled,
     |   ^
  36 |   headerBackTitle,
  37 |   headerBackTitleStyle,
  38 |   headerBackTitleVisible = true,

Call Stack
  HeaderConfig (my-app/node_modules/@react-navigation/native-stack/src/views/HeaderConfig.tsx:35:3)
  React.forwardRef$argument_0 (my-app/node_modules/react-native/Libraries/Animated/createAnimatedComponent.js:39:6)
  ExpoRoot (my-app/node_modules/expo-router/src/ExpoRoot.tsx:55:11)
  WithDevTools (my-app/node_modules/expo/build/launch/withDevTools.js:31:27)
  React.forwardRef$argument_0 (my-app/node_modules/react-native/Libraries/Components/View/View.js:36:7)
  React.forwardRef$argument_0 (my-app/node_modules/react-native/Libraries/Components/View/View.js:36:7)
  AppContainer (my-app/node_modules/react-native/Libraries/ReactNative/AppContainer.js:40:19)
  getCachedComponentWithDisplayName (my-app/node_modules/react-native/Libraries/ReactNative/getCachedComponentWithDebugName.js:28:80)
"
`;
